### PR TITLE
[#1713] Improvement: Don't log some errors on exit

### PR DESCRIPTION
**Pull request by EnricoMi**
### What changes were proposed in this pull request?

Once `ShuffleWriteClientImpl.close()` / `RegisterHeartBeat.shutdown()` is called, failing async calls should not be logged as `WARN` log level as those calls are expected to fail in that situation.

### Why are the changes needed?

Some async calls may not have finished on close / shutdown, which raises warnings in the log. This pollutes logs and makes tracing unrelated issues harder:

```
WARN impl.ShuffleWriteClientImpl: Error happened when unregistering to ShuffleServerInfo{host[10.109.9.45], grpc port[19999]}
io.grpc.StatusRuntimeException: CANCELLED: Thread interrupted
        at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:268)
        at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:249)
        at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:167)
        at org.apache.uniffle.proto.ShuffleServerGrpc$ShuffleServerBlockingStub.unregisterShuffleByAppId(ShuffleServerGrpc.java:772)
        at org.apache.uniffle.client.impl.grpc.ShuffleServerGrpcClient.doUnregisterShuffleByAppId(ShuffleServerGrpcClient.java:345)
        at org.apache.uniffle.client.impl.grpc.ShuffleServerGrpcClient.unregisterShuffleByAppId(ShuffleServerGrpcClient.java:352)
        at org.apache.uniffle.client.impl.ShuffleWriteClientImpl.lambda$unregisterShuffle$28(ShuffleWriteClientImpl.java:1020)
        at org.apache.uniffle.common.util.ThreadUtils.lambda$executeTasks$0(ThreadUtils.java:110)
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.InterruptedException
        at io.grpc.stub.ClientCalls$ThreadlessExecutor.throwIfInterrupted(ClientCalls.java:750)
        at io.grpc.stub.ClientCalls$ThreadlessExecutor.waitAndDrain(ClientCalls.java:718)
        at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:159)
        ... 9 more
```
Fix: #1713

### Does this PR introduce _any_ user-facing change?

Yes, log messages are more specific on the situation when async calls fail after close / shutdown has been called.

### How was this patch tested?
No.

### CODE REVIEW

File: client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java [hunk 1]
```diff
@@ -914,7 +915,16 @@ public void sendAppHeartbeat(String appId, long timeoutMs) {
               LOG.warn("Failed to send heartbeat to " + shuffleServerInfo);
             }
           } catch (Exception e) {
-            LOG.warn("Error happened when send heartbeat to " + shuffleServerInfo, e);
+            if (closed) {
+              // this client has been closed, so failing heartbeat calls are expected
+              LOG.info(
+                  "Error happened when send heartbeat to "
+                      + shuffleServerInfo
```
Comment (rickyma) [inline client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java:922]: It's better to use  `{}` placeholders.

File: client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java [hunk 2]
```diff
@@ -914,7 +915,16 @@ public void sendAppHeartbeat(String appId, long timeoutMs) {
...
+                      + " after shuffle write client has been closed",
+                  e);
+            } else {
+              LOG.warn("Error happened when send heartbeat to " + shuffleServerInfo, e);
```
Comment (rickyma) [inline client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java:926]: Ditto.

File: client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java (hunk 1 above)
Comment (rickyma) [reply client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java:922]: Also, `when sending` is better.

File: server/src/main/java/org/apache/uniffle/server/RegisterHeartBeat.java [hunk 3]
```diff
@@ -137,7 +140,9 @@ boolean sendHeartBeat(
               sendSuccessfully.set(true);
             }
           } catch (Exception e) {
-            LOG.error(e.getMessage());
+            if (!shutdown) {
+              LOG.error(e.getMessage());
```
Comment (rickyma) [inline server/src/main/java/org/apache/uniffle/server/RegisterHeartBeat.java:144]: I think `LOG.error("xxx", e);` is better than `LOG.error(e.getMessage());`. Because it will give a full stack trace.

File: client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java (hunk 2 above)
Comment (rickyma) [reply client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java:926]: For other similar logging issues in this PR, please modify them together as well. :)

**Comment by rickyma** [comment]
BTW, I think `Error happened when unregistering to ShuffleServerInfo` error has nothing to do with `ShuffleWriteClientImpl.close() / RegisterHeartBeat.shutdown()` being called. Because unregistering threads are in a different thread pool which will not be shutdown by calling `ShuffleWriteClientImpl.close() / RegisterHeartBeat.shutdown()`. It might be shutdown when `spark.rss.client.unregister.request.timeout.sec` timeout is reached.

**Comment by EnricoMi** [comment]
You are right @rickyma. An executor finishes by calling close / shutdown and then may terminate. That interrupts waiting threads and causes ongoing calls to throw exceptions. Since close / shutdown has been called, we don't want those exceptions to be logged as errors / warns.

There are two options
1) wait gracefully for the open requests to complete or timeout, but this defers the executor termination, which might not be desired
2) terminate the executor quickly while relying on clean-up mechanisms (current behaviour)

We don't care about the heartbeats, and shuffles are cleaned-up after some timeout anyway.

File: client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java (hunk 2 above)
Comment (rickyma) [reply client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java]: There are still a lot of places using `+ shuffleServerInfo`, you can search for them.

**Comment by rickyma** [comment]
Do you know the real reason behind the failure of unregistering shuffles? I'm actually quite curious about this. It certainly isn't because of `ShuffleWriteClientImpl.close() / RegisterHeartBeat.shutdown()` being called. Because this would cause the shuffle to be delayed in being cleaned up on the server side, resulting in a waste of resources.

**Comment by EnricoMi** [comment]
> Do you know the real reason behind the failure of unregistering shuffles?

I think it is because the executor JVM is shutting down. It calls `ShuffleManager.stop`, which does not wait on ongoing calls to terminate, as they run in a different thread pool.

**Comment by rickyma** [comment]
> I think it is because the executor JVM is shutting down. That might interrupt the waiting threads (`InterruptedException`).

Yeah, what improvement measures can we take to avoid this situation as much as possible?

**Comment by EnricoMi** [comment]
Waiting for the queue of requests to complete. Is the that desired behaviour?

**Comment by rickyma** [comment]
Normally, the executor should shutdown gracefully. `InterruptedException` is not a desired behavior.

**Comment by rickyma** [comment]
> @rickyma what about #1736?

The idea looks good to me.

**Comment by EnricoMi** [comment]
I have checked the implementation of `ThreadUtils.executeTasks`. That method is meant to return **after** all tasks (unregister requests) have completed, either successfully, with failure, or after timeout. Then, the executor is shutdown so that any ongoing request is being cancelled.

So I am not sure why we are seeing those interrupt exceptions. Either these indicate timeouts or some other cause interrupt.

The `unregisterRequestTimeSec` refers to invoking all unregister requests, not individual requests. If `unregisterThreadPoolSize` is smaller than `shuffleServerInfos.size()`, then some requests are not executed in parallel.

**Comment by rickyma** [comment]
As you said https://github.com/apache/incubator-uniffle/pull/1714#issuecomment-2120490454, it could be JVM shutting down which causes `InterruptedException`. I think this [patch](https://github.com/apache/incubator-uniffle/pull/1736) might work, could you please test it?

**Comment by EnricoMi** [comment]
> As you said [#1714 (comment)](https://github.com/apache/incubator-uniffle/pull/1714#issuecomment-2120490454), it could be JVM shutting down which causes `InterruptedException`. I think this [patch](https://github.com/apache/incubator-uniffle/pull/1736) might work, could you please test it?

By the time `close` is called, all threads are completed. Calling `unregisterExecutorService.awaitTermination()` then is redundant. I'll check if the observed issue is due to too many shuffle servers with to few threads.

**Comment by rickyma** [comment]
> By the time `close` is called, all threads are completed. 

If I'm correct, this is not guaranteed. These are different ExecutorServices.

**Comment by EnricoMi** [comment]
> > By the time `close` is called, all threads are completed.
> 
> If I'm correct, this is not guaranteed. These are different ExecutorServices.

Let me rephrase this: By the time `ShuffleWriteClientImpl.unregisterShuffle` returns, all unregister requests completed.

**Comment by EnricoMi** [comment]
I think the right thing to do here is to use the timeout for each individual request and to wait for the concurrent requests according to their _sequenciality_ (i.e. the number of sequential requests given the number of threads): #1738

**Comment by rickyma** [comment]
> Let me rephrase this: By the time `ShuffleWriteClientImpl.unregisterShuffle` returns, all unregister requests completed.

This is right. The method is not asychronized.

**Comment by EnricoMi** [comment]
This is a genuine problem, so do not ignore the error. Proper fix is #1766 and #1738.
